### PR TITLE
Pickle WorkerState

### PR DESCRIPTION
**Pull request by crusaderky**
Provide initial support for pickling the WorkerState.
This is not (yet) used for dumping; consensus should be reached before we do anything about it.
For the time being, this is necessary because of https://github.com/dask/dask-benchmarks/pull/50.

### CODE REVIEW

File: distributed/tests/test_collections.py [hunk 1]
```diff
@@ -148,3 +153,29 @@ def __init__(self, i):
...
+
+    Note
+    ----
+    To make this test work with plain pickle and not need cloudpickle, we had to avoid
+    lambdas and local classes in our test. Here we're testing that HeapSet doesn't add
+    lambdas etc. of its own.
+    """
+    heap = HeapSet(key=operator.attrgetter("i"))
+
+    # Test edge case with broken weakrefs
```
Comment (crusaderky) [inline distributed/tests/test_collections.py:169]:
```suggestion
    # The heap contains broken weakrefs
```

File: distributed/tests/test_worker_state_machine.py [hunk 2]
```diff
@@ -152,6 +153,27 @@ def test_WorkerState__to_dict(ws):
...
 
+def test_WorkerState_pickle(ws):
+    """Test pickle round-trip.
+
+    Big caveat
+    ----------
+    WorkerState, on its own, can be serialized with pickle; it doesn't need cloudpickle.
+    A WorkerState extracted from a Worker might, as data contents may only be
+    serializable with cloudpickle. Some objects created externally - namely, the
+    SpillBuffer - may not be serializable at all.
```
Comment (crusaderky) [inline distributed/tests/test_worker_state_machine.py:164]:
```suggestion
    serializable with cloudpickle. Some objects created externally and not designed for
    network transfer - namely, the SpillBuffer - may not be serializable at all.
```

File: distributed/worker_state_machine.py [hunk 3]
```diff
@@ -1062,7 +1061,7 @@ def __init__(
         self.has_what = defaultdict(set)
         self.data_needed = HeapSet(key=operator.attrgetter("priority"))
         self.data_needed_per_worker = defaultdict(
-            lambda: HeapSet(key=operator.attrgetter("priority"))
+            cast(Callable, partial(HeapSet, key=operator.attrgetter("priority")))
```
Comment (graingert) [inline distributed/worker_state_machine.py:1064]: this should work without a cast https://mypy-play.net/?mypy=latest&python=3.10&flags=strict&gist=e7185bc44629717e7af27fe9d359a046

```suggestion
            partial(HeapSet[TaskState], key=operator.attrgetter("priority"))
```

File: distributed/collections.py [hunk 4]
```diff
@@ -53,6 +53,22 @@ def __init__(self, *, key: Callable[[T], Any]):
     def __repr__(self) -> str:
         return f"<{type(self).__name__}: {len(self)} items>"
 
+    def __reduce__(self) -> tuple[Callable, tuple]:
+        heap = [(k, i, v) for k, i, vref in self._heap if (v := vref()) in self._data]
```
Comment (graingert) [inline distributed/collections.py:57]: is the worker state ever pickled from a different thread than where it's being mutated?

```suggestion
        heap = [(k, i, v) for k, i, vref in self._heap.copy() if (v := vref()) in self._data]
```

Comment (fjetter) [reply distributed/collections.py:57]: unlikely but not impossible.

Standard pattern (beyond the current benchmarkign stuff in https://github.com/dask/dask-benchmarks/pull/50)

would be something like `client.run` or a newer version of `client.dump_cluster_state` which would all be executing on the main htead.

However, we cannot protect from improper usages, e.g.

```python
def task():
    return pickle.dumps(get_worker().state)
client.submit(task)
```

is valid dask but would pickle in another thread. I don't think we should protect or guard against this if it is expensive/ `_heap` is a list here, isn't it? a shallow copy should be fine but I'm not convinced it is necessary.

File: distributed/tests/test_worker_state_machine.py [hunk 5]
```diff
@@ -152,6 +153,27 @@ def test_WorkerState__to_dict(ws):
...
+    serializable with cloudpickle. Some objects created externally and not designed
+    for network transfer - namely, the SpillBuffer - may not be serializable at all.
+    """
+    ws.handle_stimulus(
+        AcquireReplicasEvent(
+            who_has={"x": ["127.0.0.1:1235"]}, nbytes={"x": 123}, stimulus_id="s1"
+        )
+    )
+    ws.handle_stimulus(UpdateDataEvent(data={"y": 123}, report=False, stimulus_id="s"))
+    ws2 = pickle.loads(pickle.dumps(ws))
```
Comment (fjetter) [inline distributed/tests/test_worker_state_machine.py:172]: that's very exciting

File: distributed/collections.py (hunk 4 above)
Comment (crusaderky) [reply distributed/collections.py:57]: None of the class is thread-safe; I don't see why pickling should be an exception.
Also, there's no guarantee that `list.copy()` will hold the GIL - if it does, it's an implementation detail.

Comment (crusaderky) [reply distributed/collections.py:57]: > We perhaps should explicitly document what non-trivial operations are atomical (for example list and dict copying is atomic) and whether atomacity is the part of the language specification or CPython implementation detail.

For as long as it's not explicitly documented, we must treat it as an implementation detail we can't rely upon

**Review by fjetter — APPROVED** [review]
